### ISSUE TITLE
Optional input validation of orchestrator.conf.json

### ISSUE THREAD

**Issue by sjmudd** [openark/orchestrator#82]
I noticed some configuration settings in my `/etc/orchestrator.conf.json` were no longer valid. Orchestrator reads in the file and does not indicate there's an issue of any sort.

While I'm happy with lazy evaluation and also potentially having invalid settings in the config file it would be most helpful I think to be able to identify such invalid settings.  Two thoughts come to mind:

### (A) command line "validate my config file" option

`$ orchestrator -c check_validate_config_file`

which would indicate if the config contents are good or not. Invalid JSON settings would generate an error.

### (B): Optional setting to fail if the current config is invalid

Something like: `"ValidateConfigSetting": true,`

This is the setting I would prefer but of course setting it now would do nothing until the option is actually implemented, so I probably need to use a combination of both (A) and (B).

Anyway this issue is to simply reflect this potential problem. There are a large number of options that can be used in the config file, a type is quite easy to make, or simply code that is changing rapidly may leave a setting which is expected to do something to be invalid and be ignored rather than generate an error.

**Comment by shlomi-noach** [comment]
Good idea. What I'd suggest though it to allow a `_` prefixed variable as "silently ignore".
So, for example:

```
{
  "NoSuchParam": 7,
  "_NoSuchParam": 8
}
```

The first would yield error, the 2nd wouldn't.

**Comment by orkestrov** [comment]
Is there a good reason to not make parameter validation mandatory? I think it is always easy to run new code against old config and correct all inconsistencies before deployment.

Over a couple years we've accumulated quite some errors due to code change and human mistakes.
```
2017/03/07 15:51:55 Unexpected config option 'ActiveNodeNotificationProcesses' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'BackendDBConcurrency' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'DenyAutoPromotionHostnamePattern' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'DiscoveryConcurrencyLoggingPeriod' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'DiscoveryPollSeconds' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'HostnameSuffixSanityHint' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'IgnoreUnreachableMasterWithStaleSlaves' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'MySQLDiscoveryReadTimeoutSeconds:' in file /etc/orchestrator.conf.json
2017/03/07 15:51:55 Unexpected config option 'MySQLTopologyReadTimeoutSeconds:' in file /etc/orchestrator.conf.json
```

**Comment by orkestrov** [comment]
Validation above was done with https://github.com/orkestrov/orchestrator/commit/782f75a998cd8e082085f84c7b4fa01a846f0f73.

**Comment by shlomi-noach** [comment]
@orkestrov please see my comment: https://github.com/github/orchestrator/issues/82#issuecomment-281379394

As long as you silently allow `_` prefix, I agree with mandatory validation.

**Comment by orkestrov** [comment]
Cool. And yes: I already ignore `_` prefixed fields.

**Comment by sjmudd** [comment]
@shlomi-noach I'd say that `_` is like MySQL's loose_ prefix. I don't like that much.
Have been discussing with @orkestrov about this.  I think that in most cases you do want to fail when there's an error in the config and not let it go unnoticed.

I'm partly of the opinion that we should fail hard if the config is wrong. Let's see if / where this causes issues (mainly on downgrades I think) and if so provide a non-fatal behaviour (via a new option) if the config parameters are not good.  ???
